packages/integration_test/lib/integration_test.dart

```diff
@@ -6,70 +6,158 @@ import 'dart:async';
+  reportTestException = (FlutterErrorDetails details, String testDescription) {
+    registerException('Test $testDescription failed: $details');
+  };
+
+  final Completer<Map<String, Object>> resultsCompleter =
+      Completer<Map<String, Object>>();
+
+  await directRunTests(
+    testMain,
+    reporterFactory: (Engine engine) =>
+        ResultReporter(engine, resultsCompleter),
+  );
+
+  final Map<String, Object> results = await resultsCompleter.future;
+
+  binding._updateTestResultState(results);
+  await reporter.report(results);
+}
+
+/// Abstract interface for a result reporter.
+abstract class Reporter {
+  /// Reports test results.
+  ///
+  /// This method will be called at the end of [run] with the [results] of
+  /// running the test suite.
+  Future<void> report(Map<String, Object> results);
+}
+
+/// Default implementation of the reporter that sends results over to the
+/// platform side.
+class _ReporterImpl implements Reporter {
+  const _ReporterImpl();
+
+  @override
+  Future<void> report(
+    Map<String, Object> results,
+  ) async {
+    try {
+      await IntegrationTestWidgetsFlutterBinding._channel.invokeMethod<void>(
+        'allTestsFinished',
+        <String, dynamic>{
+          'results': <String, String>{
+            for (final MapEntry<String, Object> result in results.entries)
+              result.key: result.value is Failure
+                  ? _formatFailureForPlatform(result.value as Failure)
+                  : result.value as String
+          }
+        },
+      );
+    } on MissingPluginException {
+      print('Warning: integration_test test plugin was not detected.');
+    }
+  }
+}
 
-const String _success = 'success';
+String _formatFailureForPlatform(Failure failure) =>
+    '${failure.error} ${failure.details}';
 
 /// A subclass of [LiveTestWidgetsFlutterBinding] that reports tests results
 /// on a channel to adapt them to native instrumentation test format.
-class IntegrationTestWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding implements IntegrationTestResults {
-  /// Sets up a listener to report that the tests are finished when everything is
-  /// torn down.
+class IntegrationTestWidgetsFlutterBinding extends LiveTestWidgetsFlutterBinding
+    implements IntegrationTestResults {
+  /// If [run] is not used, sets up a listener to report that the tests are
+  /// finished when everything is torn down.
+  ///
+  /// This functionality is deprecated – clients are expected to use [run] to
+  /// execute their tests instead.
   IntegrationTestWidgetsFlutterBinding() {
-    // TODO(jackson): Report test results as they arrive
+    if (!_isUsingLegacyReporting) {
+      // TODO(jiahaog): Point users to use the CLI https://github.com/flutter/flutter/issues/66264.
+      print('Using the legacy test result reporter, which will not catch all '
+          'errors thrown in declared tests. Consider wrapping tests with [run] '
+          'instead.');
```

Comment:
Rather than `[run]` here, I think we probably want a link to either API docs or a wiki page about this.

Reply:
Also, now that this lives in the Flutter repo, we may want the `flutter drive` command to handle this - /cc @jonahwilliams for thoughts on that.

Reply:
I'm not sure the context, but we could accept more command line args for the drive script if that is what you mean?

Reply:
@jonahwilliams There's more detailed context in https://github.com/flutter/flutter/issues/66264.

The short summary is that `flutter drive` is a pretty low level CLI that is not too friendly for users when they need to specify different tests / driver scripts. In addition, there's a bit of boilerplate that can we can hide from users in some sort of executable that we provide. With this PR, users will need to wrap their tests with run

Before

```dart
import 'package:integration_test/integration_test.dart';

void main() {
  testWidgets("failing test example", (WidgetTester tester) async { ... });
}
```

After

```dart
import 'package:integration_test/integration_test.dart';

void main() => run(_testMain);

void _testMain() {
  testWidgets("failing test example", (WidgetTester tester) async { ... });
}
```

It will be great if we could hide the need to call `run` from users, in a "bootstrap" sort of functionality in the executable – I believe `package:flutter_test` and `package:test` do something similar in the executables that are provided.

Reply:
Ahh okay. No I think putting this in integration_test is the right way to go. The tool API should remain as is  - and generally speaking we should avoid increasing that surface area because it will be difficult/impossible to change

Reply:
I'm not familiar with the flutter release process, will having a separate tool for integration tests in `packages/integration_test/bin/` be the correct approach here? (Users will be invoking it with something like `flutter pub run integration_test`)